### PR TITLE
[Live] Max Bid UI support

### PR DESCRIPTION
**Pull request by orta**
Will inline comment, some of this stuff is complex, then I can figure what to do WRT actual code-comments.

No tests for the new stuff yet

### CODE REVIEW

File: Artsy/Networking/Live_Auctions/LiveAuctionSocketCommunicator.swift [hunk 1]
```diff
@@ -127,7 +127,8 @@ private extension SocketSetup {
         case "OperationFailedEvent": break;
             // TODO: Handle op failure
 
-        case "PostEventResponse":
+
+        case "CommandSuccessful", "CommandFailed", "PostEventResponse":
```
Comment (orta) [inline Artsy/Networking/Live_Auctions/LiveAuctionSocketCommunicator.swift]: OK, these are the _responses_ to the request to put in a max bid:

```
Received socket event type: CommandSuccessful.
Received unknown socket event type. Payload: ["wasAccepted": 1, "key": F4F07E8F-A16F-4C42-861D-58CB5A0F0FD2, "event": {
    amountCents = 750000;
    bidder =     {
        bidderId = 940216;
        type = ArtsyBidder;
    };
    createdAt = "2016-06-04T12:19:48.669Z";
    eventId = "91959ee3-a9b7-47a8-b225-8c0e483adf96";
    lotId = 54c7eff17261692b51460200;
    type = SecondPriceBidPlaced;
}, "type": CommandSuccessful]
Received socket event type: LotUpdateBroadcast.
Event: ["amountCents": 500000, "bidder": {
    bidderId = 940216;
    type = ArtsyBidder;
}, "createdAt": 2016-06-04T12:19:48.669Z, "eventId": 91959ee3-a9b7-47a8-b225-8c0e483adf96-0, "lotId": 54c7eff17261692b51460200, "type": FirstPriceBidPlaced]
```

Note: raising a max bid doesn't work yet on causality, so  you get a fail when you try do it a second time

```
Received socket event type: CommandFailed.
Received unknown socket event type. Payload: ["type": CommandFailed, "wasAccepted": 0, "key": 19304909-9AD6-4108-A3EF-A0B020DAC498, "event": {
    amountCents = 750000;
    bidder =     {
        bidderId = 940216;
        type = ArtsyBidder;
    };
    createdAt = "2016-06-04T12:19:55.025Z";
    eventId = "0822b761-d00d-437f-bd1b-ecdb8e4a0ccb";
    lotId = 54c7eff17261692b51460200;
    type = SecondPriceBidPlaced;
}, "reason": {
    type = BidderIsAlreadyWinning;
}]
```

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionBidButton.swift [hunk 2]
```diff
@@ -24,7 +24,11 @@ func ==(lhs: LiveAuctionBidButtonState, rhs: LiveAuctionBidButtonState) -> Bool
 class LiveAuctionBidButton : ARFlatButton {
     var viewModel: LiveAuctionBiddingViewModelType
     var outbidNoticeDuration: NSTimeInterval = 1
-    
+
+    // On the lotVC we want to indicate being outbid
+    // but on the max-bid modal we don't.
+    var flashOutbidOnBiddableStateChanges = true
```
Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionBidButton.swift]: the bid button derives the outbid state on a lot by seeing a change from max -> biddable, but on a max bid modal this isn't wanted.

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionBidButton.swift [hunk 3]
```diff
@@ -144,13 +148,17 @@ class LiveAuctionBidButton : ARFlatButton {
                 let formattedPrice = price.convertToDollarString(currencySymbol)
                 handleBiddable(buttonState, formattedPrice: formattedPrice)
 
-            case .BiddingInProgress:
+            case .BiddingInProgress, .BidConfirmed:
```
Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionBidButton.swift]: `BidConfirmed` in this case is that the server said "I know this bid exists" not that it is the top bidder, maybe `BidAcknowledged`is better

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionBidHistoryViewController.swift [hunk 4]
```diff
@@ -94,7 +94,7 @@ class LiveAuctionBidHistoryViewController: UITableViewController {
 
         // We may be _re_appearing, so scroll to top if we have any cells.
         if lotViewModel.numberOfDerivedEvents > 0 {
-            tableView.scrollToRowAtIndexPath(NSIndexPath(forRow: 0, inSection: 0), atScrollPosition: .Top, animated: false)
+//            tableView.scrollToRowAtIndexPath(NSIndexPath(forRow: 0, inSection: 0), atScrollPosition: .Top, animated: false)
```
Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionBidHistoryViewController.swift]: this is crashing for me when I close the modal

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionPlaceMaxBidViewController.swift [hunk 5]
```diff
@@ -5,7 +5,9 @@ enum LiveAuctionBiddingProgressState {
     case TrialUser
     case Biddable(askingPrice: UInt64, currencySymbol: String)
     case BiddingInProgress
-    case BidSuccess
+    case BidConfirmed
+    case BidBecameMaxBidder
+    case BidOutbid
```
Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionPlaceMaxBidViewController.swift]: I debated for a while about the state here, but I feel like a the singular state like `NewMaxBid(isMaxBidder: Bool)` felt weirder than separating these two out

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionPlaceMaxBidViewController.swift [hunk 6]
```diff
@@ -143,6 +149,8 @@ class LiveAuctionPlaceMaxBidViewController: UIViewController {
         // New bids can come in while we are on this screen
         bidViewModel.currentBid = max(bidViewModel.currentBid, bidViewModel.currentLotValue)
         updateBiddingControls(bidViewModel.currentBid)
+
+        shouldShowBiddingOverlay(false, maxBidder: bidViewModel.lotViewModel.userIsHighestBidder)
```
Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionPlaceMaxBidViewController.swift]: normal lot events trigger a hide for the overlay, and say whether the user is the max bidder

Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionPlaceMaxBidViewController.swift]: a bid progress, and bid confirmed ones send in a nil for the max bidder later in the file, so we know that we don't know yet

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionStateManager.swift [hunk 7]
```diff
@@ -67,7 +67,30 @@ class LiveAuctionStateManager: NSObject {
...
+//                        setupWithState(.Active(biddingState: .BidNetworkFail))
+//                    } else {
+//
+//                        if success {
+//                            setupWithState(.Active(biddingState: .BidSuccess))
+//                        } else {
+//                            self.setupUI("Outbid", background: red)
+//                        }
+//                    }
+//                }
```
Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionStateManager.swift]: this can go now

**Comment by ArtsyMobile** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Error">
        
            1 Error
        
      </th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="true">There were <a href="https://eigen-ci.s3.amazonaws.com/2758/index.html">snapshot errors</a></td>
    </tr>
    
    
  </tbody>
</table>
  

  
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Warning">
        
            2 Warnings
        
      </th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:warning:</td>
      <td data-sticky="true">Needs testing on a Phone if change is non-trivial</td>
    </tr>
        <tr>
      <td>:warning:</td>
      <td data-sticky="true">Detected some Swift building time outliers</td>
    </tr>
    
    
  </tbody>
</table>
  

  



Time | Class | Function |
| --- | ----- | ----- |246.0ms | [AuctionViewController.swift](/artsy/eigen/tree/orta-max_bid/Artsy/View_Controllers/Auction/AuctionViewController.swift#L181) | @objc dynamic func showRefineTappedAnimated(animated: Bool)
199.5ms | [SwiftyJSON.swift](/artsy/eigen/tree/orta-max_bid/Pods/SwiftyJSON/Source/SwiftyJSON.swift#L445) | get {}
173.2ms | [WebSocket.swift](/artsy/eigen/tree/orta-max_bid/Pods/Starscream/Source/WebSocket.swift#L528) | private func processRawMessage(buffer: UnsafePointer<UInt8>, bufferLen: Int)
91.7ms | [Observable.swift](/artsy/eigen/tree/orta-max_bid/Pods/Interstellar/Sources/Observable.swift#L71) | private final func nextTokenHash() -> Int
71.6ms | [LiveAuctionsAdminViewController.swift](/artsy/eigen/tree/orta-max_bid/Artsy/View_Controllers/Live_Auctions/LiveAuctionsAdminViewController.swift#L22) | @objc override func viewDidLoad()
69.5ms | [AuctionLotMetadataStackScrollView.swift](/artsy/eigen/tree/orta-max_bid/Artsy/View_Controllers/Live_Auctions/Views/AuctionLotMetadataStackScrollView.swift#L18) | required init(viewModel: LiveAuctionLotViewModelType, sideMargin: String)
59.7ms | [RefinementOptionsViewController+Private.swift](/artsy/eigen/tree/orta-max_bid/Artsy/View_Controllers/Auction/RefinementOptionsViewController+Private.swift#L63) | func stackView() -> ORStackView




<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://github.com/danger/danger/">danger</a>
</p>

**Comment by ashfurrow** [comment]
💯

**Pull request merged**